USB: HID: Add new callback and refactor sample

finikorg opened the pull request.

finikorg committed:
usb: hid: Add status_cb to hid ops

Adding status callback allows to control report sending only when i.e.
device is connected or configured.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: hid: General code cleanup

Trivial style cleanup.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: hid: Refactor basic HID sample

Refactor USB writes, start transfer only when device
is configured, the change makes it possible to work with native_posix
USB virtual controller.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

aurel32 commented on samples/subsys/usb/hid/src/main.c:
```diff
@@ -81,19 +83,45 @@ int get_report_cb(struct usb_setup_packet *setup, s32_t *len,
+		     USB_TRANS_WRITE | USB_TRANS_NO_ZLP,
+		     hid_write_cb, NULL);
+}
+
```
What is the goal of using `usb_transfer` here instead of using `hid_int_ep_write` directly from `in_ready_cb`. Your solution might work, but I think the goal of a sample is to show how to use the existing API.

finikorg replied:
@aurel32 Ok, fixed to using HID API

aurel32 commented on samples/subsys/usb/hid/src/main.c:
```diff
@@ -81,19 +83,45 @@ int get_report_cb(struct usb_setup_packet *setup, s32_t *len,
+	switch (status) {
+	case USB_DC_CONFIGURED:
+		hid_write_cb(0x81, 0, NULL);
+	default:
```
In that case you can just call `in_ready_cb` instead to start the transfer loop.

aurel32 reviewed (commented):
The first two commits looks like nice improvement, it's nice being able to get the control status report from the application.

I agree the sample should be change to only start the transfer once the device is configured, but I think that as a sample it should show how to use the HID API, not the underlying USB API.

finikorg committed:
samples: hid: Add README for the sample

Describe building and testing of the sample.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: hid-mouse: Trivial README correct

Correct title marking.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

pdunaj commented on include/usb/class/usb_hid.h:
```diff
@@ -54,6 +54,7 @@ struct usb_hid_descriptor {
 typedef int (*hid_cb_t)(struct usb_setup_packet *setup, s32_t *len,
 			u8_t **data);
 typedef void (*hid_int_ready_callback)(void);
+typedef void (*status_cb_t)(enum usb_dc_status_code status, u8_t *param);
```
can memory pointed by param change in the callback? if not pointer should be const

finikorg replied:
I took this from 
```
typedef void (*usb_status_callback)(enum usb_dc_status_code status_code,
				    u8_t *param);
```
in `usb_device.h`

finikorg replied:
I think we can make it const though even in the stack

finikorg replied:
I have fixed this

pdunaj commented on samples/subsys/usb/hid/src/main.c:
```diff
@@ -53,6 +53,8 @@ static const u8_t hid_report_desc[] = {
 
+static u8_t report_1[2] = { REPORT_ID_1, 0x00 };
+
 int debug_cb(struct usb_setup_packet *setup, s32_t *len,
```
it is not under this change but can you make all these callbacks static?

finikorg replied:
done

aurel32 commented:
Thanks @finikorg for the changes. I have looked at it and the code seems all fine now. However I have tried it on a real board (nucleo_l432kc) and now that the calls to `hid_int_ep_write` are chained to the `in_ready_cb` instead of being a regular call, the loop stops after 15-20 times. I haven't debugged it further, it might be a bug in the USB driver. It might be worth trying on another board.

dbkinder commented on samples/subsys/usb/hid/README.rst:
```diff
@@ -0,0 +1,55 @@
+********
+
+This sample app demonstrates use of a USB Human Interface Device (HID) driver
+by the Zephyr project.  This very simple driver enumerated as a raw HID device.
```
change to ``driver is enumerated``

dbkinder commented on samples/subsys/usb/hid/README.rst:
```diff
@@ -0,0 +1,55 @@
+
+After you have built and flashed the sample app image to your board, plug the
+board into a host device, for example, a PC running Linux.
+The board will be detected as shown by the Linux dmesg command:
```
You wrote you were using the ``dmesg`` command, but you used ``journalctl`` in your example below.

finikorg committed:
samples: hid: Make callbacks static

Make callbacks static according to standards.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: hid: Use delayed work for report send

Delegate report sending to delayed work.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
@aurel32 I tested this with quark_se and it was working but now I added delayed_work report sender with 2 sec delay, it would be the same behaviour as before.

finikorg committed:
tests: Fix sanity_check

Use right order for tests.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

aurel32 commented on samples/subsys/usb/hid/src/main.c:
```diff
@@ -81,36 +85,55 @@ int get_report_cb(struct usb_setup_packet *setup, s32_t *len,
+	switch (status) {
+	case USB_DC_CONFIGURED:
+		in_ready_cb();
+	default:
```
A break is missing there, causing the following line to be printed:
```
[main] [WRN] status_cb: status 3 unhandled
```

finikorg replied:
My bad, fixed now.

aurel32 reviewed (commented):
Thanks @finikorg for the new version. I have tested it on my STM32 board and it now works all fine. I guess before the MCU was overloaded by the speed of the debug logs.

I have just noticed a missing `break` in the switch, but besides that it looks all good to me.

finikorg committed:
usb: Use always single callback definition

Use one callback type, remove duplicates.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Use const for status_callback parameter

usb_dc_status_callback() parameters are interface or configuration
numbers and should be const.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jfischer-no reviewed (approved):
LGTM

aurel32 reviewed (approved):
Thanks for the new version, it looks all good.

finikorg commented:
@dbkinder @pdunaj Could you recheck the changes I made.

pdunaj commented on samples/subsys/usb/hid/src/main.c:
```diff
@@ -17,6 +17,10 @@
 #define REPORT_ID_1	0x01
 #define REPORT_ID_2	0x02
 
+struct k_delayed_work delayed_report_send;
```
static

pdunaj commented on samples/subsys/usb/hid/src/main.c:
```diff
@@ -81,36 +85,57 @@ int get_report_cb(struct usb_setup_packet *setup, s32_t *len,
+
+static void in_ready_cb(void)
+{
+	k_delayed_work_submit(&delayed_report_send, REPORT_TIMEOUT);
```
Can it happen that this function will be called again before REPORT_TIMEOUT time passes?

dbkinder commented on samples/subsys/usb/hid/README.rst:
```diff
@@ -0,0 +1,55 @@
+Requirements
+************
+
+This project requires an USB device driver.
```
Sorry I missed this earlier.  Change “an USB” to “a USB”

finikorg replied:
@dbkinder Could you recheck?

dbkinder reviewed (changes requested):
One doc change

dbkinder reviewed (approved):
LGTM, thanks!

nashif commented:
@finikorg rebase please

nashif commented:
maybe you want to move the sample to the new logger now that USB has moved..

finikorg committed:
samples: hid: Use new logger

Use new USB logger with function names.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
@nashif updated HID sample to the new USB logger
